checkmate/models/data/custom_rule.py

```diff
@@ -0,0 +1,36 @@
+"""Model for our own blocking rules."""
+
+import sqlalchemy as sa
+from sqlalchemy.dialects.postgresql import ARRAY
+
+from checkmate.checker.url.reason import Reason
+from checkmate.db import BASE
+
+
+class CustomRule(BASE):
+    """Rule about blocking a particular resource."""
+
+    __tablename__ = "custom_rule"
+    __table_args__ = (sa.Index("ix__hash", "hash"),)
+
+    id = sa.Column(sa.Integer, autoincrement=True, primary_key=True)
+
+    # While our hashes should be unique, we might change our mind about how
+    # we hash. The rule itself should stay the same
+    rule = sa.Column(sa.String, nullable=False, unique=True)
+    """The text of the rule"""
+
+    # We'd like to know if we have a hash collision
+    # The C collation here allows the B-Tree default indexing type to be used
+    # for prefix matching. This makes searching for 'ab2d23d45a%' very quick
+    hash = sa.Column(sa.String(collation="C"), nullable=False, unique=True)
```

Comment:
The actual tip off is here: https://www.postgresql.org/docs/9.5/indexes-types.html and not very obvious:

> The optimizer can also use a B-tree index for queries involving the pattern matching operators LIKE and ~ if the pattern is a constant and is anchored to the beginning of the string — for example, col LIKE 'foo%' or col ~ '^foo', but not col LIKE '%bar'. However, if your database does not use the C locale you will need to create the index with a special operator class to support indexing of pattern-matching queries; see Section 11.9 below. 

I'm not really sure what a "special operator class" is, but this seems like the other option.

This might actually not work for Google WebRisk annoyingly as part of the update process requires you to sort and hash the hashes to check if you're up to date. If the ordering is different from their expected ordering this won't work. As this is hex, I suspect it might be the same? Guess we'll find out.

Reply:
Done!